cilium/CFP-32759-cilium-mtls-via-mtls-agent.md

```diff
@@ -0,0 +1,260 @@
+
+When such a packet is seen, we check node B’s BPF map for the destination pod’s IP. As Pod B is mTLS-enabled, its IP will be a key in the map. We then look up mTLS Agent B by the well-known key, and again, rewrite the packet’s destination MAC address to that of mTLS Agent B and redirect it to the appropriate network interface.
+
+When the packet reaches mTLS Agent B’s pod network namespace, TPROXY rules in the mTLS agent network namespace will match the packet based on its destination port 15008 and redirect it to 127.0.0.1:15008 via the local loopback device. The packet will reach the process listening on the mTLS Agent at port 15008.
+
+![alt text](images/cilium-mTLS-datapath-step-2.png)
+
+#### Step 3: From mTLS Agent B to Pod B
+
+When the packet reaches mTLS Agent B, we are able to read the request’s header and get the original destination port. If not already established, a new TCP session is established with Pod B at the original destination port. The packet is then routed from mTLS Agent B to Pod B using the standard routing rules of the cluster.
+
+![alt text](images/cilium-mTLS-datapath-step-3.png)
+
+## Requirements
+
+The targeted Cilium release for an MVP of mTLS via transparent proxying is v1.17. 
+
+#### MVP Components
+1. Transparent Redirection
+1. Cilium xDS Server 
+1. mTLS via Transparent Proxy
+
+### MVP Scope
+
+Users can enable Cilium mTLS per workload. All traffic leaving a node from an mTLS-enabled source destined for an mTLS-enabled destination on a different node will be sent over mTLS.
+
+This proposal is scoped to the following traffic configurations:
+* Traffic between workloads on the same node will not be proxied (and therefore not be encrypted)
+* Cross-node traffic on a single cluster is tunneled over mTLS
+* Node egress traffic to non-mTLS-enabled destinations is not proxied
+
+Out of scope configurations (design should be extensible to support these configuration, but they are out of scope for this proposal):
+* Multi-cluster traffic
+
+### Enablement API
+
+For an MVP, the mTLS agent functionality will be enabled in Cilium via an environment variable in the cilium configuration - _mtlsAgent: enable_. The Cilium agent will identify the mTLS agent via a pod label - _cilium.io/redir-type=proxy_. Workloads are enabled for redirection and identified by the Cilium agent with the following label - _cilium.io/redir-type=workload_. For the MVP, the user will apply the labels per workload. A future goal is to support enabling workloads by namespace with support for exclusions. The Cilium agent will use these labels to identify which pods should be added to the BPF redirection map and to validate requests for workload configuration.
```

Comment:
As I mentioned on our call today, I think that enablement should behave like enablement does for the current mutual auth solution:

* we have options for the _existing_ mutual auth and encryption settings that select this behavior. `ztunnel` would be an acceptable name for both to me.
* traffic is selected for mutual auth/mTLS using the existing mechanism inside using the existing `authentication.mode: required` config inside the exsiting CiliumNetworkPolicy and CiliumClusterwideNetworkPolicy objects.
* Because this existing mechanism allows the selection of traffic using labels on the Pods, it can allow opting-in on a per-Pod (or per workload) level.
* However, namespace labels are currently not supported, so we may need to leave that for a later date.